### ISSUE TITLE
Cannot add working infinitude instances to HA

### ISSUE THREAD

**Issue by dustins5523** [MizterB/homeassistant-infinitude-beyond#33]
I have 2 thermostats, each have their own raspberry pi running portainer and their own image of infinitude. I can visit the IP addresses and use infinitude properly in everyway. And the old home assistant integration infinitude used to work, but stopped. Since then I've been using the cloud based integration. But since I have the infinitudes running I want to install infinitude beyond. Every time I put in the IP address it either comes back with "unexpected error" or just freezes and forces HA to reboot. I'm baffled as to why a previously working system has stopped working with no changes that I can remember. What am I doing wrong?

**Comment by dustins5523** [comment]
I completely re did my infinitude set up, fresh install. The thermostat AND infinitude are working but infinitude (the old integration) nor this integration will connect. I have unblocked the ports on my firewall on the docker host. I have disabled the carrier infinity cloud component, several reboots ect. No matter how i format the IP it will not connect. If I put only the IP (192.168.1.120 without http://) it will get stuck and reboot HA. If include http:// it give unexpected error. Please help
![Screenshot 2024-11-01 221250](https://github.com/user-attachments/assets/02483e0e-d58e-49a4-9478-89484555248a)

**Comment by greggitter** [comment]
Since no one replied I'll ask, you verified you're using the IP address for infinitude and not the thermostat (in the screenshot above)?  Is HA on a different subnet than infinitude (since you mentioned a firewall rule).  Can you ping the infinitude server from the HA server?  And of course, you are using the latest version of Inifinitude?  The author doesn't create version numbers for that software, so you may need to manually check for recent updates.  Admittedly these are the simple things to check.  Good luck.

**Comment by dustins5523** [comment]
Thank you for responding to this issue thats making me pull my hair out!

The IP above is the IP for Infinitude, not the thermostat. I have 2 docker containers running, one for each of my thermostats. 192.168.1.120:3000 and 192.168.1.120:3001 I cant access either of them with HA. I can view the web interface of infnitiude and do everything as expected there. My HA instance is at 192.1268.1.3 so same subnet as well. 

I had a working infinitude system working for 2 years maybe, then all the sudden things just sopped working and I had to move to the cloud integration, which i dont like. But since then Ive completely redone my infinitude installs, these are maybe a week old right now, so they are updated and working as expected.

I do suspect maybe the problem is on the HA side, but have no idea how to diagnose or fix? I dont know how to ping from HA to infinitude, can you help with that?

I mentioned the firewall rules because my docker is on a windows machine that has a plex server and a bunch of other stuff on it, and to get somethings working on that I have to create firewall rules for those ports because windows was blocking it, so I have those ports open both directions. And its not a windows issue because my previous working infinitudes were on raspberry pis.

edit:

I also have all of my home network traffic running through AdGuard. I have tried disabling that, as well as configuring the tstats directly to go to adguard. But when I had it working a long time ago I also had adguard, so I dont think its that

BUT if I could learn how to determine if HA can ping infinitude that might be helpful. Thanks again for helping

**Comment by greggitter** [comment]
Our setups are at opposite ends of the spectrum.  I don't think I'm going to be much help on this but maybe a couple things.

First, to ping your docker instance from HA you would need to open an SSH session on your HA hardware and enter "ping 192.168.1.120".  If it replies then you know it exists and have at least some connectivity.  Also, have you tried stopping one of the containers and seeing if you can add just one infinitude?

**Comment by dustins5523** [comment]
Thank you again for your tips!

I can ping to both the IP and the port, so we are good there. I Shutdown one container and tried and it still doesnt connect. I did finally check my HA logs and I get 2 errors when I try to install it, maybe this can help

First one:
This error originated from a custom integration.

Logger: custom_components.infinitude_beyond.infinitude.api
Source: custom_components/infinitude_beyond/infinitude/api.py:75
integration: Infinitude Beyond (documentation)
First occurred: November 2, 2024 at 7:51:38 PM (20 occurrences)
Last logged: 1:04:33 PM

Cannot connect to host http:80 ssl:default [None]
Cannot connect to host http:443 ssl:default [None]

Second one:
This error originated from a custom integration.

Logger: custom_components.infinitude_beyond.config_flow
Source: custom_components/infinitude_beyond/config_flow.py:64
integration: Infinitude Beyond (documentation)
First occurred: November 2, 2024 at 7:51:38 PM (5 occurrences)
Last logged: 1:04:33 PM

Unexpected exception
Traceback (most recent call last):
  File "/config/custom_components/infinitude_beyond/config_flow.py", line 64, in async_step_user
    await validate_input(self.hass, user_input)
  File "/config/custom_components/infinitude_beyond/config_flow.py", line 43, in validate_input
    await infinitude.connect()
  File "/config/custom_components/infinitude_beyond/infinitude/api.py", line 174, in connect
    status, config, energy, profile = await asyncio.gather(
                                      ^^^^^^^^^^^^^^^^^^^^^
  File "/config/custom_components/infinitude_beyond/infinitude/api.py", line 152, in _fetch_config
    data = resp.get("data", {})
           ^^^^^^^^
AttributeError: 'NoneType' object has no attribute 'get'

**Comment by dustins5523** [comment]
I am not using SSL on anything, nor is the box ticked, FYI

**Comment by greggitter** [comment]
You might try adding to allow windows firewall to access ports 80 and 443 on your docker instance.  Also, ping doesn't ping specific ports, just the IP...ping is ICMP it's own port.  Those lines of code are indeed where infinitude is trying to connect to your docker instance....something appears to be blocking it.

Edit:  Of course the firewall rule you add would be incoming from your HA ip address.  Probably just need port 80.  Looking closer at the code it does seem to be specific port 3000 (as the default) but maybe testing it with that other port will provide an additional clue.  Here's a snippet from infinitude/api.py.

```
     30 class Infinitude:
     31     """Object for interacting with the Infinitude API."""
     32
     33     def __init__(
     34         self,
     35         host: str,
     36         port: int = 3000,
     37         ssl: bool = False,
     38         *,
     39         session: Optional[ClientSession] = None,
     40     ) -> None:
     41         """Initialize the Infinitude API."""
     42         self.host: str = host
     43         self.port: str = port
     44         self.ssl = ssl
     45         self._session: ClientSession = session
     46
     47         self._status: dict = {}
     48         self._config: dict = {}
     49         self._energy: dict = {}
     50         self._profile: dict = {}
     51
     52         if not self._session or self._session.closed:
     53             self._session = ClientSession()
     54
     55         self.system: InfinitudeSystem
     56         self.zones: dict[int, InfinitudeZone]
     57
     58     @property
     59     def url(self):
     60         """Get the base URL to connect to Infinitude."""
     61         protocol = "http"
     62         if self.ssl:
     63             protocol = "https"
     64         return f"{protocol}://{self.host}:{self.port}"
     65
     66     async def _get(self, endpoint: str, **kwargs) -> dict:
     67         """Make a GET request to Infinitude."""
     68         url = f"{self.url}{endpoint}"
     69         try:
     70             async with self._session.get(url, **kwargs) as resp:
     71                 data: dict = await resp.json(content_type=None)
     72                 resp.raise_for_status()
     73                 return data
     74         except ClientError as e:
     75             _LOGGER.error(e)
```

**Comment by dustins5523** [comment]
I went overboard on the firewall rules and have allowed any connection on 80, 443, 3000, and 3001 on TCP and UDP incoming and outgoing and nothing. I have even gone back to trying the address without the http:// and it just times out and causes HA to reboot

My server can ping HA and HA can ping the server. What the hell am I doing wrong

**Comment by dustins5523** [comment]
I just got a new error, maybe this will help:

This error originated from a custom integration.

Logger: custom_components.infinitude_beyond.infinitude.api
Source: custom_components/infinitude_beyond/infinitude/api.py:75
integration: Infinitude Beyond (documentation)
First occurred: 9:17:07 PM (4 occurrences)
Last logged: 9:17:07 PM

Cannot connect to host http:80 ssl:default [DNS server returned answer with no data]

**Comment by dustins5523** [comment]
The plot thickens.

I installed a new test Home assistant OS on my windows computer using virtual machine and I was able to connector both thermostats no problem. It does not work when you use http:// but when you type in just the IP address and port they both connect.

The HA ive been using has been working well on a RPi 3 for many years. Maybe its time for a complete fresh install? I dont know, but the plug in is working, and my dockers and network are working. For purposes of maybe helping someone one day I will reply soon after I completely redo my Rpi, come up with a more permanent solution. THANK YOU for your help and tips, you motivated me to keep looking into it!

**Comment by MizterB** [comment]
I can confirm that including the http:// in the host field is incorrect, and is reflected in the error log where is thinks that ‘http’ is the host name.

**Comment by greggitter** [comment]
> THANK YOU for your help and tips, you motivated me to keep looking into it!

Not sure if I helped at all but glad you got it figured out...or are on your way.  I'm confused what actually changed.  You said you installed a new HA version on Windows but the other version is on you rpi...?  But that was an older version?  Still seems like a connectivity issue...since you moved it to the same hardware the infinitude is running did that maybe solve the connection issue?  Either way glad you have a working solution.  Cheers.

**Comment by dustins5523** [comment]
So I have an fully running HA setup running Home assistant OS on a Rpi3. My docker containers are on a windows server. 

Last night on my personal PC I put together a home assistant OS VM on virtual box, set it up, and tried to set up infinitude and it worked. The docker instance and the VM are on 2 different PCs. 

What I think it means is that it's not a network issue, or a docker issue. Maybe the rpi is timing out and rebooting? Maybe the SD card is corrupted? I don't know yet. My next step is going to be installing a fresh ha os on a new SD card on that same rpi and setup infinitude first and see if it works. Idk, maybe it's something in the code causing my ha to crash. But in this very same pi, same as card I have the original infinitude integration working and without me making any changes one day it just was broken and never worked again, and that led me to the new integration and my struggles. 

So I'll post my results with a new card and fresh install.

**Comment by dustins5523** [comment]
I bought a Home Assistant Green to upgrade my RPi 3. The HA Green is way faster than the RPi, I used a fresh install, no back up. The UI is much faster, installing HACS and all that is super fast. But when I go to setup Infinitude Beyond it just gets stuck. Now incredibly slow to reboot. There is something in the code that is causing this, it has to be. Why would it install in a snap on a HA OS docker container, but break a brand new HA Green? 

So frustrating when this is THE reason I bought new hardware. As I learn more, I will post

**Comment by 7ooL** [comment]
I found this updated version today and have been using infinitude for years. I had no issues adding two instances that run in containers on the same IP with different ports. It was straightforward with using just the IP address and port number. 

I was able to remove the older yaml configurations and uninstall the older infinitude repository. HA made it easy to rename the systems and zones to the old names allowing previous automations to work without edit (Spook help to highlight possible issues). Thank you @MizterB for this project.
